enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,260 @@
+The bootstrap node creates a temporary control plane that is required for launching the actual cluster.
+
+Resource constraints are the primary motivation for single-node OCP clusters, and removing the need for a
+separate bootstrap machine significantly reduces the resources required to install the cluster.
+
+Requiring the auxiliary node for installing a Single Node OpenShift brings to play some requirements:
+1. The obvious additional node.
+2. Requires external dependencies:
+   a. Load balancer (only for bootstrap phase)
+   b. Requires DNS (configured per installation)
+3. Cannot use Bare Metal IPI:
+   a. Adds irrelevant dependencies - vips, keepalived, mdns
+   b. Requires same L2 between bootstrap and the Single Node OpenShift
+
+### Goals
+
+* This enhancement describes an approach for installing Single Node OpenShift in a BareMetal environment
+  for production use.
+* Minimal changes to openShift installer, this implementation shouldn't effect existing deployment flows.
+* Installation should result a clean Single Node OpenShift, without bootstrap leftovers.
+* Rely on a livecd, so it would be able to trigger the full installation using virtual media.
+* Self-contained bootstrap, The bootstrap flow is expressed as a single Ignition config which is layered on top of a pristine RHCOS image. It's important that we limit our use of this pattern to just the Ignition configs - as opposed to including other assets directly in the ISO image. 
+
+### Non-Goals
+
+* This enhancement does not address similar installation flow for 3 nodes cluster.
+* This enhancement does not address high-availability for single-node deployments.
+* This enhancement does not address single-node-developer cluster-profile installation.
+* This enhancement does not address cloud deployment for bootstrap in place (since livecd cannot be used). It will be addressed as part of future enhancement.
+* This enhancement does not address updating single-node clusters. It will be addressed as part of future enhancement.
+
+## Proposal
+
+Execute the bootstrap flow on a liveCD.
+Enrich the master Ignition with the control plane static pods and write this Ignition.
+Write the updated master Ignition + RHCOS to disk and reboot.
+Post reboot the node will complete the installation.
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+Ignition configuration.
+The user will boot RHCOS liveCD with this Ignition to create a single-node deployment.
+
+### Implementation Details/Notes/Constraints
+
+The OpenShift installer `create ignition-configs` command will generate a `bootstrap-in-place-for-live-iso.ign`
+file, when the number of replicas for the control plane (in the `install-config.yaml`) is `1`.
+
+The `bootstrap-in-place-for-live-iso.ign` will be embedded into an RHCOS liveCD using coreos-install embed command.
```

Comment:
Who will do this? The user? Or an orchestration tool like Hive/ACM?

Reply:
Yeah, who is the actor here? Mainly wondering if it's the installer if these components get embedded into openshift-install? are they even written in Go?

Reply:
This came up in another doc...I think basically we have a product gap around scripting the ISO, [the docs say it's "not supported"](https://docs.openshift.com/container-platform/4.6/installing/installing_bare_metal/installing-bare-metal.html#installation-user-infra-machines-advanced_ignition_installing-bare-metal) but the plan is to fix this with something like https://github.com/openshift/enhancements/pull/492

Actually the more I think about this the more I feel like we may end up wanting a new:
`openshift-install create-single-node-iso` that can also take things like additional manifests for setting the RT kernel (and kernel args) via MachineConfig as well as supporting injecting network configuration as files and choosing the target disk drive for `coreos-installer`.

Something like this:

```
$ cat manifests/rt.yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker"
  name: worker-kerneltype
spec:
  kernelType: realtime
$ cat single-node-iso-target.yaml
# Gets passed to coreos-installer on Live ISO
coreos-install:
  device: /dev/nvme0n1
  copy-network: true
# Gets written in the Live ISO, then copied to system via above
NetworkManager:
  vlan-eth0.conf: |
    [connection]
    id=VLAN for building 4A
    uuid=8ce1c9e0-ce7a-4d2c-aa28-077dda09dd7e
    interface-name=VLAN-4A
    type=vlan

    [vlan]
    interface-name=VLAN-4A
    parent=eth0
    id=4
$ openshift-install create single-node-iso --manifests=manifests/ --nodeconf=single-node-iso-target.yaml
```

Reply:
Internally, `create single-node-iso` would compile the `single-node-iso-target.yaml` into Ignition (much like https://github.com/coreos/fcct/issues/137 plus the `NetworkManager` sugar I just made up) and include it along with the Ignition it generates and inject it into the ISO.

Reply:
+ it should also propagate some of the parameters to the coreos-install on the pivot phase as well. (https://github.com/openshift/installer/pull/4482/files#diff-f6cfc5d463e8139ffa14ab1b1b448280664b32bfb20b60cef2479e27561d6a33R389)

Reply:
Anyhow, I would say it's out of scope for this enhancement and should be submitted as a separate one later on. WDYT?

Reply:
Yes...we can defer a "high level" interface but I think we should settle on the "low level" interface.  Which as noted elsewhere it might be enough to offer a `SNO_COREOS_INSTALLER_ARGS=` environment variable but...honestly I think it wouldn't be very much more code to support an interface (like what fcct does) to add arbitrary files into the live environment (which would include `coreos-installer` arguments but also potentially e.g. NetworkManager configs which I am pretty sure are going to be required in some circumstances like a small virtual media ISO).

Reply:
I think must of this is covered under https://github.com/openshift/enhancements/pull/565/files#diff-f323c28b80ad16321f0d3b7775dd2aeba57289409b516b6aaae0362fe3337ebfR215
About HIVE/ACM, to my knowledge, they will integrate with assisted-installer which support this flow (it's delivering the ignition to a node that runs with liveCD and orchestrate the installation).

Reply:
About the interface for supporting arbitrary configuration to the live environment.
As you mentioned it's enabled by fcct and I think adding this capability to the installer can be part of future work.
`SNO_COREOS_INSTALLER_ARGS=` environment variable allows the basic configuration and seems good enough to start with.